File: py/utils/history.py
"""A module to manage the history of the user's input."""

from utils.observable import Observable


class HistoryManager(Observable):
    """A class to manage the history of the user's input."""

    def __init__(self) -> None:
        super().__init__()
        self._history = []
        self._future = []

    def __str__(self) -> str:
        return f"HistoryManager({self._history}, {self._future})"

    @property
    def history(self):
        """Get the history."""
        return list(self._history)

    @property
    def future(self):
        """Get the future."""
        return list(self._future)

    def add(self, item):
        """Add an item to the history."""
        self._history.append(item)
        self._future = []
        self._call()

    def undo(self):
        """Undo the last action."""
        if not self._history:
            return False
        self._future.insert(0, self._history.pop())
        self._call()
        return True

    def redo(self):
        """Redo the last action."""
        if not self._future:
            return False
        self._history.append(self._future.pop(0))
        self._call()
        return True

    def clear(self):
        """Clear the history."""
        self._history = []
        self._future = []
        self._call()

    def __iter__(self):
        return iter(self._history)

    def __getitem__(self, item):
        return self._history[item]

File: src/app/utils/history.ts
import { SelfObservable } from "./observable";

export default class HistoryManager<T> extends SelfObservable<
  HistoryManager<T>
> {
  private _history: T[] = [];
  private _future: T[] = [];
  private _searchIdx: number = 0;

  constructor() {
    super();
  }

  get past() {
    return this._history.slice();
  }

  get latest(): T | undefined {
    return this._history[this._history.length - 1];
  }

  get future() {
    return this._future.slice();
  }

  set past(past: T[]) {
    this._history = past;
    this._call();
  }

  toString() {
    return `HistoryManager(${this._history}, ${this._future})`;
  }

  add(item: T) {
    this._searchIdx = 0;
    this._history.push(item);
    this._future = [];
    this._call();
  }

  undo() {
    this._searchIdx = 0;
    if (this._history.length === 0) {
      return undefined;
    }
    const last = this._history.pop()!;
    this._future.unshift(last);
    this._call();
    return last;
  }

  redo() {
    this._searchIdx = 0;
    if (this._future.length === 0) {
      return undefined;
    }
    const next = this._future.shift()!;
    this._history.push(next);
    this._call();
    return next;
  }

  clear() {
    this._searchIdx = 0;
    this._history = [];
    this._future = [];
    this._call();
  }

  get(idx: number) {
    return this._history[idx];
  }

  [Symbol.iterator]() {
    return this._history[Symbol.iterator]();
  }

  get search() {
    return {
      backward: () => {
        if (this._searchIdx < this.past.length) {
          this._searchIdx++;
          return this.past[this.past.length - this._searchIdx];
        }
        return undefined;
      },
      forward: () => {
        if (this._searchIdx > 0) {
          this._searchIdx--;
          return this.past[this.past.length - this._searchIdx];
        }
        return undefined;
      },
    };
  }
}
